client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -0,0 +1,306 @@
+                try {
+                    onClose.run();
+                    socket.close();
+                } catch (IOException e1) {
+                    log.warn("Failed to close socket while failing.", e1);
+                }
+                throw Exceptions.sneakyThrow(e);
+            }
+        }, executor);
+    }
+
+    private static TrustManagerFactory createFromCert(String trustStoreFilePath)
+            throws CertificateException, IOException, NoSuchAlgorithmException, KeyStoreException {
+        TrustManagerFactory factory = null;
+        if (!Strings.isNullOrEmpty(trustStoreFilePath)) {
+            KeyStore trustStore = CertificateUtils.createTrustStore(trustStoreFilePath);
+
+            factory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            factory.init(trustStore);
+        }
+        return factory;
+    }
+
+    /**
+     * Creates a socket connected to the provided endpoint. 
+     * Note that this is a sync call even though it is called in an async context. 
+     * While this is normally frowned upon, it is simply not possible to construct an SSL socket asynchronously in Java.
+     */
+    @SneakyThrows //Is called inside a completable future.
+    private static Socket createClientSocket(PravegaNodeUri location, ClientConfig clientConfig) {        
+        Socket result;
+        if (clientConfig.isEnableTlsToSegmentStore()) {
+            TrustManagerFactory trustMgrFactory = createFromCert(clientConfig.getTrustStore());
+
+            // Prepare a TLS context that uses the trust manager
+            SSLContext tlsContext = SSLContext.getInstance("TLS");
+            tlsContext.init(null,
+                    trustMgrFactory != null ? trustMgrFactory.getTrustManagers() : null,
+                    null);
+
+            SSLSocket s = (SSLSocket) tlsContext.getSocketFactory().createSocket();
+            // SSLSocket does not perform hostname verification by default. So, we must explicitly enable it.
+            if (clientConfig.isValidateHostName()) {
+                SSLParameters tlsParams = new SSLParameters();
+                tlsParams.setEndpointIdentificationAlgorithm("HTTPS");
+                s.setSSLParameters(tlsParams);
+            }
+            result = s;
+
+        } else {
+            result = new Socket();
+        }
+        result.setSendBufferSize(TCP_BUFFER_SIZE);
+        result.setReceiveBufferSize(TCP_BUFFER_SIZE);
+        result.setTcpNoDelay(true);
+        result.connect(new InetSocketAddress(location.getEndpoint(), location.getPort()));
+        return result;
+    }
+
+    @Override
+    public void send(WireCommand cmd) throws ConnectionFailedException {
+        if (closed.get()) {
+            throw new ConnectionFailedException("Connection is closed");
+        }
+        try {
+            encoder.write(cmd);
+        } catch (IOException e) {
+            log.warn("Error writing to connection");
```

Comment:
We should at least log the message of this exception (i.e. `e.toString()`) so that we know a bit more about why it failed.

Below too.

Reply:
Done